From 5882dfe74b33c78715fd6ac9f09a363340b7fed9 Mon Sep 17 00:00:00 2001
From: Chao Guo <chao.guo@nxp.com>
Date: Wed, 11 Sep 2024 17:19:14 +0900
Subject: [PATCH 84/91] LF-13361 gldownload: add alignment in allocation
 bufferpool

To meet the alignment requirement of GPU driver, add alignment
for bufferpool with some special sizes when gldownload propose
allocation.

Besides, in order for downstream elements to receive the correct
buffer size and stride, prioritize using the input buffer videometa
to set the output videometa.

Upstream-Status: Inappropriate [i.MX specific]

Signed-off-by: Chao Guo <chao.guo@nxp.com>
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 ext/gl/gstgldownloadelement.c    | 28 ++++++++++++++++++++++++++++
 gst-libs/gst/gl/gstglmemorydma.c | 25 ++++++++++++++++++++++---
 2 files changed, 50 insertions(+), 3 deletions(-)

diff --git a/ext/gl/gstgldownloadelement.c b/ext/gl/gstgldownloadelement.c
index 99fffc9..03c52b0 100644
--- a/ext/gl/gstgldownloadelement.c
+++ b/ext/gl/gstgldownloadelement.c
@@ -829,6 +829,12 @@ static void gst_gl_download_element_finalize (GObject * object);
 #define EXTRA_CAPS_TEMPLATE2
 #endif
 
+#if GST_GL_HAVE_VIV_DIRECTVIV
+#define DEFAULT_ALIGN 16
+#else
+#define DEFAULT_ALIGN 32
+#endif
+
 static GstStaticPadTemplate gst_gl_download_element_src_pad_template =
     GST_STATIC_PAD_TEMPLATE ("src",
     GST_PAD_SRC,
@@ -1423,6 +1429,7 @@ gst_gl_download_element_propose_allocation (GstBaseTransform * bt,
   GstStructure *config;
   gsize size;
   GstVideoFormat fmt;
+  gboolean need_alignment = FALSE;
 
   gst_query_parse_allocation (query, &caps, NULL);
   if (!gst_video_info_from_caps (&info, caps)) {
@@ -1439,6 +1446,8 @@ gst_gl_download_element_propose_allocation (GstBaseTransform * bt,
 #if GST_GL_HAVE_IONDMA || GST_GL_HAVE_DMABUFHEAPS
   if (fmt == GST_VIDEO_FORMAT_RGBA || fmt == GST_VIDEO_FORMAT_RGB16) {
     allocator = gst_gl_memory_dma_allocator_obtain ();
+    if (allocator)
+      need_alignment = TRUE;
     GST_DEBUG_OBJECT (bt, "obtain dma memory allocator %p.", allocator);
   }
 #endif
@@ -1478,6 +1487,25 @@ gst_gl_download_element_propose_allocation (GstBaseTransform * bt,
   gst_buffer_pool_config_add_option (config,
       GST_BUFFER_POOL_OPTION_GL_SYNC_META);
 
+  if (need_alignment) {
+    guint width, height;
+    GstVideoAlignment alignment;
+    width = GST_VIDEO_INFO_WIDTH (&info);
+    height = GST_VIDEO_INFO_HEIGHT (&info);
+
+    // add alignment in config when using dma allocator
+    memset (&alignment, 0, sizeof (GstVideoAlignment));
+    alignment.padding_right = GST_ROUND_UP_N (width, DEFAULT_ALIGN) - width;
+    alignment.padding_bottom = GST_ROUND_UP_N (height, DEFAULT_ALIGN) - height;
+    gst_buffer_pool_config_add_option (config,
+        GST_BUFFER_POOL_OPTION_VIDEO_ALIGNMENT);
+    gst_buffer_pool_config_set_video_alignment (config, &alignment);
+
+    // add video meta option to get alignment from buffer.
+    gst_buffer_pool_config_add_option (config,
+        GST_BUFFER_POOL_OPTION_VIDEO_META);
+  }
+
   //set allocator to buffer pool
   if (allocator) {
     gst_buffer_pool_config_set_allocator (config, allocator, &params);
diff --git a/gst-libs/gst/gl/gstglmemorydma.c b/gst-libs/gst/gl/gstglmemorydma.c
index 3160cb2..ceee2af 100644
--- a/gst-libs/gst/gl/gstglmemorydma.c
+++ b/gst-libs/gst/gl/gstglmemorydma.c
@@ -257,19 +257,38 @@ gst_gl_memory_dma_buffer_to_gstbuffer (GstGLContext * ctx, GstVideoInfo * info,
 {
   GstBuffer *buf;
   GstGLMemoryDMA *glmem;
+  GstVideoMeta *vmeta;
+  GstVideoMeta *ometa;
+  guint width = 0, height = 0;
+  const gsize *offsets = NULL;
+  const gint *strides = NULL;
 
   gst_gl_context_thread_add (ctx, (GstGLContextThreadFunc) _finish_texture,
       NULL);
 
+  vmeta = gst_buffer_get_video_meta (glbuf);
+  if (vmeta) {
+    width = vmeta->width;
+    height = vmeta->height;
+    offsets = vmeta->offset;
+    strides = vmeta->stride;
+  } else {
+    width = GST_VIDEO_INFO_WIDTH (info);
+    height = GST_VIDEO_INFO_HEIGHT (info);
+    offsets = info->offset;
+    strides = info->stride;
+  }
+
   glmem = (GstGLMemoryDMA *) gst_buffer_peek_memory (glbuf, 0);
 
   buf = gst_buffer_new ();
   gst_buffer_append_memory (buf, (GstMemory *) glmem->dma);
   gst_memory_ref ((GstMemory *) glmem->dma);
 
-  gst_buffer_add_video_meta_full (buf, 0,
-      GST_VIDEO_INFO_FORMAT (info), GST_VIDEO_INFO_WIDTH (info),
-      GST_VIDEO_INFO_HEIGHT (info), 1, info->offset, info->stride);
+  ometa = gst_buffer_add_video_meta_full (buf, 0,
+      GST_VIDEO_INFO_FORMAT (info), width, height, 1, offsets, strides);
+  if (vmeta)
+    gst_video_meta_set_alignment (ometa, vmeta->alignment);
   GST_BUFFER_FLAGS (buf) = GST_BUFFER_FLAGS (glbuf);
   GST_BUFFER_PTS (buf) = GST_BUFFER_PTS (glbuf);
   GST_BUFFER_DTS (buf) = GST_BUFFER_DTS (glbuf);
-- 
2.43.0

